File: 2022_python/220330.py
#
# created by z on 3/30/2022
# 74

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rn: int = len(matrix)
        cn: int = len(matrix[0])

        l: int = 0
        r: int = rn * cn - 1

        while l <= r:
            m: int = int((l + r) / 2)
            # transform
            row: int = m // cn
            col: int = m % cn
            # judge
            x: int = matrix[row][col]
            if x < target:
                l = m + 1
            elif x > target:
                r = m - 1
            else:
                return True

        return False

File: 2022/2022_03_cpp/220330.cpp
//
// Created by z on 3/30/2022.
// # 78
// O(log n + log m)... maybe
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rn = matrix.size(), cn = matrix[0].size();
        int l = 0, r = rn * cn - 1;

        while (l <= r)
        {
            int m = (l+r) / 2;
            int x = matrix[m/cn][m%cn];
            if (x > target)
                r = m-1;
            else if (x < target)
                l = m+1;
            else
                return 1;
        }

        return 0;
    }
};